spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectReplyToMessageListenerContainer.java

```diff
@@ -0,0 +1,198 @@
+		super.setChannelAwareMessageListener((message, channel) -> {
+			this.inUseConsumerChannels.remove(channel);
+			messageListener.onMessage(message, channel);
+		});
+	}
+
+	@Override
+	public void setMessageListener(MessageListener messageListener) {
+		super.setChannelAwareMessageListener((message, channel) -> {
+			this.inUseConsumerChannels.remove(channel);
+			messageListener.onMessage(message);
+		});
+	}
+
+	@Override
+	protected void doStart() throws Exception {
+		if (!isRunning()) {
+			this.consumerCount = 0;
+			super.setConsumersPerQueue(0);
+			super.doStart();
+		}
+	}
+
+	@Override
+	protected void processMonitorTask() {
+		long now = System.currentTimeMillis();
+		synchronized (this.consumersMonitor) {
+			long reduce = this.consumers.stream()
+				.filter(c -> this.whenUsed.containsKey(c) && !this.inUseConsumerChannels.containsValue(c)
+						&& this.whenUsed.get(c) < now - getIdleEventInterval())
+				.count();
+			if (reduce > 0) {
+				if (logger.isDebugEnabled()) {
+					logger.debug("Reducing idle consumes by " + reduce);
+				}
+				this.consumerCount = (int) Math.max(0, this.consumerCount - reduce);
+				super.setConsumersPerQueue(this.consumerCount);
+			}
+		}
+	}
+
+	@Override
+	protected void consumerRemoved(SimpleConsumer consumer) {
+		this.inUseConsumerChannels.remove(consumer.getChannel());
+		this.whenUsed.remove(consumer);
+	}
+
+	/**
+	 * Get the channel associated with a direct reply-to consumer.
+	 * @return the channel.
+	 */
+	public Channel getChannel() {
+		synchronized (this.consumersMonitor) {
+			Channel channel = null;
+			while (channel == null) {
+				if (!isRunning()) {
+					throw new IllegalStateException("Direct reply-to container is not running");
+				}
+				for (SimpleConsumer consumer : this.consumers) {
+					Channel candidate = consumer.getChannel();
+					if (candidate.isOpen() && this.inUseConsumerChannels.putIfAbsent(candidate, consumer) == null) {
+						channel = candidate;
+						this.whenUsed.put(consumer, System.currentTimeMillis());
+						break;
+					}
+				}
+				if (channel == null) {
+					this.consumerCount++;
+					super.setConsumersPerQueue(this.consumerCount);
+				}
+			}
+			return channel;
+		}
+	}
+
+	/**
+	 * Release the consumer associated with the channel for reuse.
+	 * Set cancelConsumer to true if the client is not prepared to handle/discard a
+	 * late arriving reply.
+	 * @param channel the channel.
+	 * @param cancelConsumer true to cancel the consumer.
+	 * @param message a message to be included in the cancel event if cancelConsumer is true.
+	 */
+	public void releaseConsumerFor(Channel channel, boolean cancelConsumer, String message) {
+		SimpleConsumer consumer = this.inUseConsumerChannels.remove(channel);
+		if (consumer != null && cancelConsumer) {
+			this.logger.error("Consumer canceled by client" + consumer);
+			consumer.cancelConsumer("Consumer " + this + " canceled due to " + message);
```

Comment:
Ah! Indeed.
So, we should ensure that `message` isn't `null`, when `cancelConsumer == true`.
I don't say that only one `cancelMessage` and check it for `null` would be good solution (although for computer it is really enough :smile: ), but at least we will provide the proper message for human being. Otherwise we can end up with the:
```
... canceled due to null
```